### DISCUSSION TITLE
Vanilla-extract + custom addon

### DISCUSSION THREAD

**Discussion by ggdaltoso** [storybookjs/storybook#26186] [Help]
### Summary

Hello, everyone 👋 

I'm on a long journey of rewriting React95 and replacing `styled-components` with `vanilla-extract`; everything is going well so far.
With the help of Storybook, I was able to rewrite all components. I'm now taking a look at every theme we have. The progress can be seen in this draft PR, if anyone is interested: https://github.com/React95/React95/pull/425

As you can see in the actual version, we have a custom panel addon in our storybook responsible for changing themes: https://react95.github.io/React95. It works flawlessly because the theme is just an [object](https://github.com/React95/React95/blob/master/packages/core/components/ThemeProvider/themes/bee.tsx) used in [ThemeProvider](https://github.com/React95/React95/blob/master/packages/core/components/ThemeProvider/ThemeProvider.tsx).

Since I'm removing `styled-components` to `vanilla-extract`, each theme will have its own `.css.ts` file which, with all properly configs, will work on the storybook as well. 

The only problem until here is I can't make the panel addon work with the new way we write themes. I tried a lot to make it work and the best implementation I've got is at [refactor/sb-theme-changer](https://github.com/React95/React95/tree/refactor/sb-theme-changer) branch.

You can check the problem, code, and everything else at StackBlitz:
https://pr.new/github.com/React95/React95/tree/refactor/sb-theme-changer

and the custom addon is located at `packages/core/.storybook/src/theme-changer`

Does anyone know why/how I can't import `.css.ts` files in a storybook addon?

### Create a reproduction

https://pr.new/github.com/React95/React95/tree/refactor/sb-theme-changer

**Comment by JReinhold** [comment]
The biggest problem for you is that the Storybook manager - and thus all addons - is using a separate build pipeline and configuration. This means that whatever you configure in your `vite.config.ts` or `viteFinal` doesn't have an impact on your addon, only the preview. There's no API to hook into the build pipeline of the manager UI as we're not comfortable supporting that level of flexibility.

This means that Vanilla Extract doesn't ever compile your `*.css.ts` files in the addon, and so they are being imported as a regular TS file. Based on the code you shared it looks like you sort of expected this, because your addon only really read the keys of the themes and not the values.
Without knowing the internals of Vanilla Extract, I'm assuming the problem is the Vanilla Extract calls you're doing in the actual theme file you're importing. Since they are top-level, they are being called even though you don't really use them in the addon. And I'd expect Vanilla Extract to hard bail on those function calls if they detect that they are being executed in a non-compiled environment like this is. I'm talking about these lines:

https://github.com/React95/React95/blob/ca0e46ca22ff1636a590f5718a7cafe3459d83a1/packages/core/components/ThemeProvider/themes/win95.css.ts#L30-L40

What you should probably do instead is maintain a list of theme names that you can import and use in your addon panel instead of trying to import the `X.css.ts` themes directly. I'd expect you'd need to maintain this list somewhere else anyways.

**Comment by ggdaltoso** [comment]
Thanks for your reply @JReinhold but just the theme keys wouldn't help me.

To give us some context, the [createTheme function, from vanilla-extract](https://vanilla-extract.style/documentation/api/create-theme/), will return a className for each theme and I actually need that value and not just the key itself.

To help us a little, I created a story for the theme changer, mimicking how the addon would work.
https://pr.new/github.com/React95/React95/tree/refactor/vanilla-extract

story at: `packages/core/stories/themes.stories.tsx`

### Video

https://github.com/storybookjs/storybook/assets/6536985/3d4d02b2-dc13-4d68-ad49-50ec0c11e165

**Reply by JReinhold** [reply]
Then the only option I can come up with is for your addon to set those theme keys (eg. `win95`) in the globals, and then your `withTheme` decorator to manually map those theme keys to the actual class name needed, as that decorator runs in a compiled environment.
